File: leetcode/300.最长递增子序列.ts
/*
 * @lc app=leetcode.cn id=300 lang=typescript
 *
 * [300] 最长递增子序列
 */

// @lc code=start
function lengthOfLIS(nums: number[]): number {
    const tmp: number[] = [];
    for (let n of nums) {
        const length = tmp.length;
        if (length === 0 || n > tmp[length-1]) {
            tmp.push(n);
            continue;
        }

        let [left, right] = [0, length-1];
        let p = right;
        while (left <= right) {
            const mid = (left+right) >> 1;
            if (tmp[mid] >= n) {
                p = mid;
                right = mid-1;
            } else left = mid+1;
        }
        tmp[p] = n;
    }
    return tmp.length;
};
// @lc code=end

File: leetcode/300.最长递增子序列.rs
/*
 * @lc app=leetcode.cn id=300 lang=rust
 *
 * [300] 最长递增子序列
 */

// @lc code=start
impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let mut tmp: Vec<i32> = Vec::new();
        for n in nums {
            let length = tmp.len();
            if length == 0 || n > tmp[length - 1] {
                tmp.push(n);
                continue;
            }

            let (mut left, mut right) = (0, length as i32 - 1);
            let mut p = right;
            while left <= right {
                let mid = (left + right) >> 1;
                if tmp[mid as usize] >= n {
                    p = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            tmp[p as usize] = n;
        }
        tmp.len() as i32
    }
}
// @lc code=end

